src/WIP.sol

```diff
@@ -0,0 +1,251 @@
+
+    // ERC-165
+    function supportsInterface(bytes4 interfaceID) external view returns (bool);
+
+    // ERC-2612
+    function permit(address owner, address spender, uint value, uint deadline, uint8 v, bytes32 r, bytes32 s) external;
+    function nonces(address owner) external view returns (uint);
+    function DOMAIN_SEPARATOR() external view returns (bytes32);
+
+    // Permit2
+    function permit2(address owner, address spender, uint amount, uint deadline, bytes calldata signature) external;
+}
+
+contract WIP is IWIP {
+    string public constant override name = "Wrapped IP";
+    string public constant override symbol = "WIP";
+    uint8 public override decimals = 18;
+
+    mapping (address => uint) public override balanceOf;
+    mapping (address => mapping (address => uint)) public override allowance;
+
+    // keccak256("Permit(address owner,address spender,uint256 value,uint256 nonce,uint256 deadline)");
+    bytes32 private constant PERMIT_TYPEHASH = 0x6e71edae12b1b97f4d1f60370fef10105fa2faae0126114a169c64845d6126c9;
+    // bytes4(keccak256("isValidSignature(bytes32,bytes)")
+    bytes4 private constant MAGICVALUE = 0x1626ba7e;
+    mapping(address => uint) public override nonces;
+
+    uint private immutable INITIAL_CHAIN_ID;
+    bytes32 private immutable INITIAL_DOMAIN_SEPARATOR;
+
+    constructor() {
+        INITIAL_CHAIN_ID = block.chainid;
+        INITIAL_DOMAIN_SEPARATOR = _computeDomainSeparator();
+    }
+
+    receive() external payable {
+        deposit();
+    }
+
+    function supportsInterface(bytes4 interfaceID) external pure override returns (bool) {
+        return
+            // ERC-165
+            interfaceID == this.supportsInterface.selector ||
+            // ERC-2612
+            interfaceID == this.permit.selector ||
+            // Permit2
+            interfaceID == this.permit2.selector;
+    }
+
+    function DOMAIN_SEPARATOR() public view override returns (bytes32) {
+        return block.chainid == INITIAL_CHAIN_ID ? INITIAL_DOMAIN_SEPARATOR : _computeDomainSeparator();
+    }
+
+    function _computeDomainSeparator() private view returns (bytes32) {
+        return keccak256(
+            abi.encode(
+                // keccak256("EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)")
+                0x8b73c3c69bb8fe3d512ecc4cf759cc79239f7b179b0ffacaa9a75d522b39400f,
+                // keccak256(bytes('Wrapped IP')),
+                0x4a24dd8304360c3edc71acded1e27d8467d787ccaeefb153eaaadce60e21753b,
+                // keccak256(bytes("1"))
+                0xc89efdaa54c0f20c7adf612882df0950f5a951637e0307cdcb4c672f298b8bc6,
+                block.chainid,
+                address(this)
+            )
+        );
+    }
+
+    function deposit() public payable {
+        balanceOf[msg.sender] += msg.value;
+        emit Deposit(msg.sender, msg.value);
+    }
+
+    function withdraw(uint value) external override {
+        balanceOf[msg.sender] -= value;
```

Comment:
Do we need to check `msg.sender`'s balance > value before this? Or am i missing something.

Reply:
It is expected, as after solidity 8.0, it will revert if there is underflow.